app/code/Magento/Store/App/Request/PathInfoProcessor.php

```diff
@@ -38,7 +38,20 @@ public function process(\Magento\Framework\App\RequestInterface $request, $pathI
             /** @var \Magento\Store\Api\Data\StoreInterface $store */
             $store = $this->storeManager->getStore($storeCode);
         } catch (NoSuchEntityException $e) {
-            return $pathInfo;
+            try {
+                /** get the store by url */
+                $requestUri = $request->getUri();
+                if (empty($requestUri)) {
+                    return $pathInfo;
+                }
+                $store = $this->storeManager->getStoreByUrl($requestUri);
```

Comment:
Implementation must rely on `StoreManagerInterface`, not on current implementation of `StoreManager`.